### PR TITLE
feat(api): Sort sessions by release timestamp [TET-7]

### PR DESCRIPTION
**Pull request by ahmedetefy**
This PR adds the ability to sort by release timestamp in sessions v2 over metrics
Essentially it:
- Runs a pre-flight query to `Release` model according to ordering of release timestamp whether ASC or DESC and gets the resulting release groups. It also adds environment filters to the pre flight query if environment filters are sent in the query param of the request
- Fetches those groups and uses them as a filter in subsequent metrics queries
- Formats the output so if groups are present in the `Release` model but not in the metrics dataset, those groups still exist but are nulled out in the output. Otherwise, it shows the results of the groups in the metrics dataset.
- Orders the output according to resulting groups from preflight query

### CODE REVIEW

File: tests/snuba/api/endpoints/test_organization_sessions.py [hunk 1]
```diff
@@ -1428,3 +1428,470 @@ def test_release_is_empty(self):
                     "totals": {"sum(session)": 1},
                 }
             ]
+
+
+@patch("sentry.api.endpoints.organization_sessions.release_health", MetricsReleaseHealthBackend())
```
Comment (ahmedetefy) [inline tests/snuba/api/endpoints/test_organization_sessions.py:1433]: I realize the tests are lengthy and could probably be refactored but honestly it will take away from the readability of the tests, and I am a fan of writing verbose tests

File: src/sentry/release_health/metrics_sessions_v2.py [hunk 2]
```diff
@@ -425,13 +435,79 @@ def run_sessions_query(
...
-    orderby = _parse_orderby(query, fields)
-    if orderby is None:
-        # We only return the top-N groups, based on the first field that is being
-        # queried, assuming that those are the most relevant to the user.
-        # In a future iteration we might expose an `orderBy` query parameter.
-        primary_metric_field = _get_primary_field(list(fields.values()), query.raw_groupby)
-        orderby = OrderBy(primary_metric_field, Direction.DESC)
+    ordered_preflight_filters: Dict[GroupByFieldName, Sequence[str]] = {}
+    try:
+        orderby = _parse_orderby(query, fields)
```
Comment (andriisoldatenko) [inline src/sentry/release_health/metrics_sessions_v2.py:440]: i know it was before, but i was wondering why don't you want to rename it to `order_by`?

File: src/sentry/release_health/metrics_sessions_v2.py [hunk 3]
```diff
@@ -425,13 +435,79 @@ def run_sessions_query(
...
+            raw_orderby = raw_orderby[1:]
+            direction = Direction.DESC
+        else:
+            direction = Direction.ASC
+
+        if raw_orderby not in PREFLIGHT_QUERY_COLUMNS:
+            raise exc
+
+        preflight_query_conditions = {
+            "orderby_field": raw_orderby,
```
Comment (andriisoldatenko) [inline src/sentry/release_health/metrics_sessions_v2.py:457]: same as above, especially cause we call other variables with `_`  in this dict.

File: src/sentry/release_health/metrics_sessions_v2.py [hunk 4]
```diff
@@ -425,13 +435,79 @@ def run_sessions_query(
...
+        # we have environment filters applied to the query, then we need to include the
+        # environment filters otherwise we might end up with metrics queries filters that do not
+        # belong to the same environment
+        environment_conditions = []
+        for condition in where:
+            preflight_query_condition = _get_filters_for_preflight_query_condition(
+                tag_name="environment", condition=condition
+            )
+            if preflight_query_condition != (None, None):
+                environment_conditions += [preflight_query_condition]
```
Comment (andriisoldatenko) [inline src/sentry/release_health/metrics_sessions_v2.py:475]: if i understand correctly this line are you trying to do something like:
```
 environment_conditions.append(preflight_query_condition)
```
?

File: src/sentry/release_health/metrics_sessions_v2.py [hunk 5]
```diff
@@ -484,19 +560,105 @@ def run_sessions_query(
...
-        ],
+        "groups": result_groups,
         "start": isoformat_z(metrics_results["start"]),
         "end": isoformat_z(metrics_results["end"]),
         "intervals": [isoformat_z(ts) for ts in metrics_results["intervals"]],
         "query": query.query,
     }
 
 
+def _reorder_result_groups_according_to_preflight_query_results(
```
Comment (andriisoldatenko) [inline src/sentry/release_health/metrics_sessions_v2.py:581]:
```suggestion
def _reorder_result_groups_by_preflight_query_results
```
wdyt?

File: src/sentry/release_health/metrics_sessions_v2.py [hunk 6]
```diff
@@ -484,19 +560,105 @@ def run_sessions_query(
...
+                "totals": {"sum(session)": 10},
+                "series": {"sum(session)": [10]},
+            },
+        ],
+    }
+    Then loop over the releases list sequentially, and rebuild the result_groups
+    array based on that order by appending to the list the values from that mapping
+    and accessing it through the key which is the group value
+    """
+    if len(ordered_preflight_filters) == 1:
```
Comment (andriisoldatenko) [inline src/sentry/release_health/metrics_sessions_v2.py:627]: what if ordered_preflight_filters > 1? 🤔

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 5 above)
Comment (jan-auer) [reply src/sentry/release_health/metrics_sessions_v2.py:581]: `_order_by_preflight`

File: src/sentry/release_health/metrics_sessions_v2.py [hunk 7]
```diff
@@ -570,6 +732,43 @@ def _transform_single_condition(
...
+    tag_name: str, condition: Union[Condition, BooleanCondition]
+) -> Tuple[Optional[Op], Optional[Set[str]]]:
+    """
+    Function that takes a tag name and a condition, and checks if that condition is for that tag
+    and if so returns a tuple of the op applied either Op.IN or Op.NOT_IN and a set of the tag
+    values
+    """
+    if isinstance(condition, Condition):
+        if condition.lhs == Column(tag_name):
+            if condition.op in [Op.EQ, Op.NEQ, Op.IN, Op.NOT_IN]:
```
Comment (andriisoldatenko) [inline src/sentry/release_health/metrics_sessions_v2.py:745]: for readability better to combine it with and IMO:
```
if isinstance(condition, Condition) and condition.lhs == Column(tag_name) and condition.op in [Op.EQ, Op.NEQ, Op.IN, Op.NOT_IN]:
```
python won't evaluate second condition if first is False.

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 5 above)
Comment (andriisoldatenko) [reply src/sentry/release_health/metrics_sessions_v2.py:581]: looks good to me too :)

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 2 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:440]: Do you mean `raw_orderby`?
tl;dr mypy complains 
Essentially `raw_orderby` is of type `str` while `orderby` is of type `OrderBy` (which is a dataclass)

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 4 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:475]: yes

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 3 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:457]: I wanted to make the distinction that this is of type `str` not of type `OrderBy`

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 5 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:581]: sure good suggestion

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 6 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:627]: There is validation against that early on in the function
https://github.com/getsentry/sentry/pull/34699/files#diff-9768b861d30daef14443451ae5e87b0352a6229bef88dfc9722fd06bad13afe1R786

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 7 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:745]: I don't agree here .. I think its actually harder to read a line with all three conditions on it

Comment (andriisoldatenko) [reply src/sentry/release_health/metrics_sessions_v2.py:745]: you can always split ands into 3 lines, but i'll leave it for you. I reread it carefully and we can't combine 3 ifs because if you have `raise InvalidParams` .... So best we can do here:
```diff
diff --git src/sentry/release_health/metrics_sessions_v2.py src/sentry/release_health/metrics_sessions_v2.py
index 5f327a6b73..0a7d0fb9d9 100644
--- src/sentry/release_health/metrics_sessions_v2.py
+++ src/sentry/release_health/metrics_sessions_v2.py
@@ -740,21 +740,18 @@ def _get_filters_for_preflight_query_condition(
     and if so returns a tuple of the op applied either Op.IN or Op.NOT_IN and a set of the tag
     values
     """
-    if isinstance(condition, Condition):
-        if condition.lhs == Column(tag_name):
-            if condition.op in [Op.EQ, Op.NEQ, Op.IN, Op.NOT_IN]:
-                filters = (
-                    {condition.rhs}
-                    if isinstance(condition.rhs, str)
-                    else {elem for elem in condition.rhs}
-                )
-                op = {Op.EQ: Op.IN, Op.IN: Op.IN, Op.NEQ: Op.NOT_IN, Op.NOT_IN: Op.NOT_IN}[
-                    condition.op
-                ]
-                return op, filters
-            raise InvalidParams(
-                f"Unable to resolve {tag_name} filter due to unsupported op {condition.op}"
+    if isinstance(condition, Condition) and condition.lhs == Column(tag_name):
+        if condition.op in [Op.EQ, Op.NEQ, Op.IN, Op.NOT_IN]:
+            filters = (
+                {condition.rhs}
+                if isinstance(condition.rhs, str)
+                else {elem for elem in condition.rhs}
             )
+            op = {Op.EQ: Op.IN, Op.IN: Op.IN, Op.NEQ: Op.NOT_IN, Op.NOT_IN: Op.NOT_IN}[condition.op]
+            return op, filters
+        raise InvalidParams(
+            f"Unable to resolve {tag_name} filter due to unsupported op {condition.op}"
+        )
```

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 4 above)
Comment (andriisoldatenko) [reply src/sentry/release_health/metrics_sessions_v2.py:475]: so my suggestion is just easy to read and based on this SO [answer](https://stackoverflow.com/questions/725782/in-python-what-is-the-difference-between-append-and) 2x faster ;)

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 6 above)
Comment (andriisoldatenko) [reply src/sentry/release_health/metrics_sessions_v2.py:627]: gotcha.

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 2 above)
Comment (andriisoldatenko) [reply src/sentry/release_health/metrics_sessions_v2.py:440]: so from what i understood, you can't name it `order_by` because of mypy right?

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 4 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:475]: Oh nice ! TIL

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 2 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:440]: Yes

Comment (andriisoldatenko) [reply src/sentry/release_health/metrics_sessions_v2.py:440]: gotcha.

File: src/sentry/release_health/metrics_sessions_v2.py (hunk 7 above)
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:745]: Agreed

File: tests/snuba/api/endpoints/test_organization_sessions.py [hunk 8]
```diff
@@ -1428,3 +1428,470 @@ def test_release_is_empty(self):
...
+            }
+        )
+        assert response.data["groups"] == [
+            {
+                "by": {"release": "1D", "session.status": None, "project": None},
+                "totals": {"sum(session)": 0},
+                "series": {"sum(session)": [0]},
+            },
+            {
+                "by": {"release": "1C", "session.status": None, "project": None},
```
Comment (jjbayer) [inline tests/snuba/api/endpoints/test_organization_sessions.py:1697]: Seeing the example response here, I'm not 100% convinced that handling the special case of subkeys being `None` is easier for a client to handle than simply omitting the entire group. On the other hand, having these groups listed is more explicit, so let's keep as-is.

**Review by jjbayer — APPROVED** [review]
Logic looks sound to me, and test coverage looks good! Just some minor comments, feel free to ignore if you disagree with them.

File: src/sentry/release_health/metrics_sessions_v2.py [hunk 9]
```diff
@@ -617,3 +827,43 @@ def _get_primary_field(fields: Sequence[Field], raw_groupby: Sequence[str]) -> M
...
+            else:
+                queryset = queryset.exclude(**environment_orm_conditions)
+
+        if direction == Direction.DESC:
+            queryset = queryset.order_by("-date_added", "-id")
+        else:
+            queryset = queryset.order_by("date_added", "id")
+
+        if limit is not None:
+            queryset = queryset[: limit - 1]
```
[the remark this answers is not in the document]
Comment (ahmedetefy) [reply src/sentry/release_health/metrics_sessions_v2.py:867]: I don't think that's true for django ORM limits.. the limit you pass is the limit passed to the query

[the remark this answers is not in the document]
Comment (jjbayer) [reply src/sentry/release_health/metrics_sessions_v2.py:867]: I might be missing the reason for the `- 1` here, do you actually want fewer results than what `limit` defines?

Example:
```
In [3]: limit =2; File.objects.filter()[:limit]
Out[3]: <BaseQuerySet [<File at 0x1157dc670: id=1>, <File at 0x1157dc070: id=2>]>
```
